22w42a 🐫

modmuss50 opened the pull request:
# 22w42a port

This PR ports Fabric API to 22w42a. There are a number of changes that affect fabric api on a lower level as well as higher level changes that will affect most modders. Big thanks to @shartte for helping with this port. Also thanks to the yarn contributors for the mappings.

Fabric loader 0.14.10 or later is required.

## Item Groups

22w42a changed the way ItemGroups work. `FabricItemGroupBuilder` has been removed and replaced with `FabricItemGroup`. The following snippet of code shows how to create an item group, and add the `TEST_ITEM` to it. This allows for much finer grain control over the ordering of your item groups.

```java
private static final ItemGroup ITEM_GROUP = new FabricItemGroup(new Identifier(MOD_ID, "test_group")) {
		@Override
		public ItemStack createIcon() {
			return new ItemStack(Items.DIAMOND);
		}

		@Override
		protected void addItems(FeatureSet enabledFeatures, Entries entries) {
			entries.add(TEST_ITEM);
		}
	};
```

If you wish to add items to vanilla's groups, or ones added via a mod you can do so with Fabric's new event. Unlike 1.19.2 you also have fine grain control over the ordering of other item groups.

```java
ItemGroupEvents.modifyEntriesEvent(ItemGroups.BUILDING_BLOCKS).register((content) -> {
		content.add(TEST_ITEM);

		// Add a diamond before the oak fence
		content.addBefore(Blocks.OAK_FENCE, Items.DIAMOND);
		// Add an emerald after the oak door.
		content.addAfter(Blocks.OAK_DOOR, Items.EMERALD);
});
```

## Biome API changes

Previously dynamic registries would source their content from both the `BuiltInRegistries` and data packs. Now the built-in registries are only used for data generation,
and content at runtime is only loaded from data packs.

Mods that previously used JSON objects for these registries exclusively are unaffected by this change. Mods that registered their worldgen content to `BuiltinRegistries` will now have
to generate JSON files and include them in their mod assets. A convenience data provider has been added (`FabricBuiltinRegistriesProvider`) 
that will automatically export JSON files for all objects in `BuiltinRegistries` belonging to the current mod. This means for mods that previously
did built-in worldgen, this should be the only required change to continue working.

The biome modification API will continue to work as before, and will throw more detailed error messages when world-gen 
objects existing only in `BuiltinRegistries` but not in data packs are used.

The following data types are affected by this:

* Dimension Types
* Biomes
* Message Types
* Configured Carvers
* Configured Features
* Placed Features
* Structures
* Structure Sets
* Structure Processor Lists
* Structure Pools
* Chunk Generator Settings
* Noise
* Density Functions
* World Presets
* Flat Level Generator Presets

## Data Generation

Minor breaking changes, recipes now require a `RecipeCategory`

# Texture API

The sprites that should be baked into texture atlases are no longer discovered by querying all `UnbakedModel` during the model baking process. Instead, the `SpriteAtlasManager` will independently search for textures using patterns in all available resource packs and bake them into specific texture atlases.

For example, `BakedModelManager` defines the block atlas as:

* All textures in the `block` folder (or any subfolder)
* All textures in the `item` folder (or any subfolder)
* All textures in the `entity/conduit` folder (or any subfolder)
* The texture at `BellBlockEntityRenderer.BELL_BODY_TEXTURE`
* The texture at `EnchantingTableBlockEntityRenderer.BOOK_TEXTURE`

In Fabric, the `ClientSpriteRegistryCallback` event has been modified to allow new rules to be appended for any texture atlas.

For example, the following would add all textures from the subfolder `part` (recursively) to the atlas (from *any* resource pack/namespace), and also the texture called "ae2:models/skyblockchest" specifically.

```java
ClientSpriteRegistryCallback.event(SpriteAtlasTexture.BLOCK_ATLAS_TEXTURE).register((resourceManager, sprites) -> {
  class_7766.method_45834(resourceManager, "part", sprites::put);
  class_7766.method_45830(resourceManager, new Identifier("ae2:models/skyblockchest"), sprites::put);
});
```


## UnbakedModel Changes

UnbakedModels no longer have texture dependencies (see Texture API), so the corresponding method was removed without a replacement.
Textures must either be in block/ or item/ to be discovered for the block atlas, or the Texture API needs to be used to add them to the atlas.

In addition, Vanilla changed how parent models are loaded by JSON models. It introduced `method_45785` to allow a model to load its parent models.
This method is only called for models that are initially added to the model load list. If your mod subclasses `UnbakedModel` and uses nested JSON models,
it should always call `method_45785` on its own model dependencies. The used JSON models would not be able to load their parents otherwise.

Example implementation:

```java
class CustomUnbakedModel implements UnbakedModel {
[...]

  @Override
  default void method_45785(Function<Identifier, UnbakedModel> loader) {
    for (Identifier dependency : getModelDependencies()) {
      loader.apply(dependency).method_45785(loader);
    }
  }
}
```

modmuss50 committed:
First look at 1.19.3

modmuss50 committed:
First go at new item group api

modmuss50 committed:
Misc fixes

shartte committed:
Port resourceloader to latest snapshot (#6)

* Port resourceloader to latest snapshot

* Port resourceloader to latest snapshot

modmuss50 committed:
Mapping update and misc fixes

modmuss50 committed:
Datagen compile pass

modmuss50 committed:
Datagen test fixes

shartte committed:
Registry Fixes (#7)

* Regsync Fixes

* Regsync Fixes

modmuss50 committed:
More compile fixes.

modmuss50 committed:
Merge remote-tracking branch 'modmuss50/1.19.3' into 1.19.3

# Conflicts:
#	fabric-biome-api-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MinecraftServerMixin.java
#	fabric-registry-sync-v0/src/testmod/java/net/fabricmc/fabric/test/registry/sync/RegistrySyncTest.java
#	settings.gradle

modmuss50 committed:
More compile fixes.

modmuss50 committed:
It compiles :)

modmuss50 committed:
Deprecated modules all just work

Gegy commented on fabric-message-api-v1/src/main/java/net/fabricmc/fabric/mixin/message/PlayerManagerMixin.java:
```diff
@@ -42,15 +42,13 @@ public abstract class PlayerManagerMixin {
-				sendMessageHeader(message, Set.of());
-			}
+			// TODO 22w42a
+//			if (!message.headerSignature().isEmpty()) {
```
These shouldn’t be needed at all anymore - the protocol is now able to safely handle gaps in players’ message chains 🙂

Gegy commented on fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/GroupResourcePack.java:
```diff
@@ -56,81 +52,58 @@ public GroupResourcePack(ResourceType type, List<ModResourcePack> packs) {
-			Collection<Identifier> modResources = pack.findResources(type, namespace, prefix, predicate);
 
-			resources.addAll(modResources);
+			pack.findResources(type, namespace, prefix, arg);
```
Not sure if it’s an issue, but question: is it guaranteed that mod packs within a group won’t have intersecting resources? Might need to filter those duplicates before invoking the consumer 🙂

LambdAurora replied:
As Gegy said, this absolutely needs filtering to de-dupe as there's no guarantee that mod resource packs won't have intersecting resources in the same group!

shartte replied:
I am not sure that vanilla does this either, however. Anyway, one more item on the big pile of "needs to be checked" things.

LambdAurora replied:
Vanilla doesn't have the dupe issue because that method exists only on resource packs that have a single root so the calls are always returning unique entries.

And ResourceManager has a very different signature, meaning when looking at resource packs it'll keep the one that has the highest priority.

shartte replied:
This method is also used in building the resource stack which returns a list of resources with the same identifier. Should that not return the asset from all mods that have it?

LambdAurora replied:
....
Damn it Mojang!

Gegy replied:
Hmm, true! It does look like stack building will work correctly when listing all resource stacks. Single resource lookups will still only yield that first entry, though- might be a weird inconsistency (also from previous behaviour).

shartte replied:
I kinda view this method as "visitResources" and the last argument as the visitor.
There are a few vanilla methods that just put the visited resources in a map. So in that case, the *last* visited resource for a given Identifier will win. We should absolutly verify: in which order does Vanilla visit the resource packs (high prio -> low prio? or reverse) and potentially at least match that deterministically (alphabetical order by mod-id maybe?)

Gegy commented on fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModNioResourcePack.java:
```diff
@@ -242,7 +236,7 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
 						try {
 							Identifier id = new Identifier(namespace, nsPath.relativize(file).toString().replace(separator, "/"));
-							if (predicate.test(id)) ids.add(id);
+							visitor.accept(id, () -> Files.newInputStream(file));
```
A sneaky change happened here: packs now need to list their mcmeta files, too

shartte replied:
Yes they do, I ran into the problem that skipping .mcmeta files in findResources actually breaks animated textures.

shartte replied:
Or are you talking about pack.mcmeta?

shartte committed:
Two more bugfixes (#8)

* Undo freezing the registries and then unfreezing again, because freezing stops intrusive holders from working in the Block registry. (Preventing additional block registrations).

* Fix registry sync packet unserializing an empty namespace without replacing it to "minecraft:".

apple502j commented on fabric-api-base/src/testmod/java/net/fabricmc/fabric/test/base/FabricApiBaseGameTest.java:
```diff
@@ -18,13 +18,11 @@
-
 public class FabricApiBaseGameTest {
-	@GameTest(templateName = FabricGameTest.EMPTY_STRUCTURE)
+	// TODO 22w42a
```
Why is this disabled?

modmuss50 replied:
Ah, had to disable to get it building when the gametest module wasnt enabled, I will fix later 👍

apple502j commented on fabric-item-group-api-v1/src/main/resources/fabric.mod.json:
```diff
@@ -16,16 +16,15 @@
   "depends": {
-    "fabricloader": ">=0.6.0",
-    "minecraft": ">=1.15-",
+    "fabricloader": ">=0.14.9",
```
```suggestion
    "fabricloader": ">=0.14.10",
```

apple502j commented on fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/mixin/registry/sync/BootstrapMixin.java:
```diff
@@ -29,6 +29,7 @@
 import net.minecraft.fluid.Fluids;
 import net.minecraft.item.Items;
 import net.minecraft.util.registry.Registry;
+import net.minecraft.util.registry.SimpleRegistry;
```
Is this import used?

apple502j commented on fabric-resource-loader-v0/src/client/java/net/fabricmc/fabric/mixin/resource/loader/client/ServerResourcePackProviderMixin.java:
```diff
@@ -0,0 +1,38 @@
+
+import net.fabricmc.fabric.impl.resource.loader.ModResourcePackCreator;
+
+@Mixin(ServerResourcePackProvider.class)
```
I feel like this should mixin to `DefaultClientResourcePackProvider`.

apple502j commented on fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/MainMixin.java:
```diff
@@ -20,15 +20,15 @@
 
 import net.fabricmc.fabric.impl.resource.loader.ModResourcePackUtil;
 
 @Mixin(Main.class)
```
I don't think this Mixin applies.

shartte replied:
This was reworked

apple502j commented on fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/TestServerMixin.java:
```diff
@@ -16,34 +16,19 @@
 
 import net.fabricmc.fabric.impl.resource.loader.ModResourcePackUtil;
 
 @Mixin(TestServer.class)
```
Don't think this is necessary.

shartte replied:
It turned out it was, but was reworked anyway. -> Enabling all resource packs like Vanilla does will enable them in the wrong order leading to Vanilla overriding mod resources.

modmuss50 committed:
Fix VibrationListenerMixin

shartte committed:
Fix checkstyle issues and fix datagen.
Also fix injection point for datapack enabling in dedicated server.
Add more validation to biome modification API due to JSON worldgen now being the norm.

shartte committed:
Fix all remaining checkstyle violations.

shartte committed:
More checkstyle fixes.
Add datagen of worldgen content to datagen-api and use it in biome api.

shartte committed:
Tests now launch.

shartte committed:
Small style fixes.

shartte committed:
Fix resource loading for TestServer.

modmuss50 committed:
Merge pull request #10 from shartte/fix-datagen

Fix datagen

modmuss50 committed:
Remove deprecated and/or broke sprite apis

shartte committed:
Fix ModelLoaderMixin.
Remove test-mod dependency on deprecated module.

shartte committed:
Change sprite registration event and wire it into the texture atlas content discovery.

shartte committed:
Javadoc

shartte committed:
Don't filter .mcmeta files because Minecraft actually needs to be able to discover these. Add a test for .mcmeta files to resource loader testmod.

modmuss50 committed:
Merge pull request #11 from shartte/mixinfix

Fix ModelLoaderMixin

modmuss50 committed:
Review fixes, and include biome api testmod generated resources.

modmuss50 committed:
Ignore cache file

modmuss50 committed:
Add IdentifiableItemGroup and MinecraftItemGroups

modmuss50 committed:
Cleanup and fixes, make MinecraftItemGroups impl.

shartte committed:
Fix invoker being captured (not respecting later event registrations).

modmuss50 committed:
Merge pull request #12 from shartte/mixinfix

Fix invoker being captured (not respecting later event registrations).

modmuss50 committed:
Start on ItemGroupEvents

modmuss50 committed:
Fix ClientPlayNetworkHandlerMixin

modmuss50 committed:
Item group fixes and tests.

modmuss50 committed:
Fix FabricItemGroupEntries interface injection.

shartte committed:
Item Group Content Modification API

shartte committed:
Entries->Content unification

shartte committed:
Refactoring (Conotent->Entries).

modmuss50 committed:
Merge pull request #13 from shartte/mixinfix

Item Group Content Modification API

modmuss50 committed:
Fix build

modmuss50 committed:
Lock supported version to 1.19.3

modmuss50 committed:
Update loom because why not.

modmuss50 commented:
This should now be ready to merge very shortly. I expect there will be stuff broken, wrong and missing but we can fix that as people find the issues.

modmuss50 committed:
Bump versions.

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+import net.minecraft.resource.featuretoggle.FeatureSet;
+
+/**
+ * This class allows the entries of {@link ItemGroup item groups} to be modified by the events in {@link ItemGroupEvents}.
```
```suggestion
 * This class allows the entries of {@link ItemGroup item groups} to be modified by the events in {@link ItemGroupEvents}.
```
```suggestion
 * This class allows the entries of {@linkplain ItemGroup item groups} to be modified by the events in {@link ItemGroupEvents}.
```

Juuxel replied:
This is still unresolved

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	/**
+	 * Adds stacks after an existing stack in the group, or at the end, if the stack isn't in the group.
+	 *
+	 * @param afterLast  Add <code>newStacks</code> after the last group entry matching this stack (compared using {@link ItemStack#canCombine}).
```
```suggestion
	 * @param afterLast  Add {@code newStacks} after the last group entry matching this stack (compared using {@link ItemStack#canCombine}).
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	 * Adds stacks after an existing stack in the group, or at the end, if the stack isn't in the group.
+	 *
+	 * @param afterLast  Add <code>newStacks</code> after the last group entry matching this stack (compared using {@link ItemStack#canCombine}).
+	 * @param newStacks  The stacks to add. Only {@link #isEnabled(ItemStack) enabled} stacks will be added.
```
```suggestion
	 * @param newStacks  The stacks to add. Only {@linkplain #isEnabled(ItemStack) enabled} stacks will be added.
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	/**
+	 * Adds stacks after an existing item in the group, or at the end, if the item isn't in the group.
+	 *
+	 * @param afterLast  Add <code>newStacks</code> after the last entry of this item in the group.
```
```suggestion
	 * @param afterLast  Add {@code newStacks} after the last entry of this item in the group.
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	 * Adds stacks after an existing item in the group, or at the end, if the item isn't in the group.
+	 *
+	 * @param afterLast  Add <code>newStacks</code> after the last entry of this item in the group.
+	 * @param newStacks  The stacks to add. Only {@link #isEnabled(ItemStack) enabled} stacks will be added.
```
```suggestion
	 * @param newStacks  The stacks to add. Only {@linkplain #isEnabled(ItemStack) enabled} stacks will be added.
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	/**
+	 * Adds stacks after the last group entry matching a predicate, or at the end, if no entries match.
+	 *
+	 * @param afterLast  Add <code>newStacks</code> after the last group entry matching this predicate.
```
```suggestion
	 * @param afterLast  Add {@code newStacks} after the last group entry matching this predicate.
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	 * Adds stacks after the last group entry matching a predicate, or at the end, if no entries match.
+	 *
+	 * @param afterLast  Add <code>newStacks</code> after the last group entry matching this predicate.
+	 * @param newStacks  The stacks to add. Only {@link #isEnabled(ItemStack) enabled} stacks will be added.
```
```suggestion
	 * @param newStacks  The stacks to add. Only {@linkplain #isEnabled(ItemStack) enabled} stacks will be added.
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	/**
+	 * Adds stacks before an existing item in the group, or at the end, if the item isn't in the group.
+	 *
+	 * @param beforeFirst Add <code>newStacks</code> before the first entry of this item in the group.
```
```suggestion
	 * @param beforeFirst Add {@code newStacks} before the first entry of this item in the group.
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	 * Adds stacks before an existing item in the group, or at the end, if the item isn't in the group.
+	 *
+	 * @param beforeFirst Add <code>newStacks</code> before the first entry of this item in the group.
+	 * @param newStacks   The stacks to add. Only {@link #isEnabled(ItemStack) enabled} stacks will be added.
```
```suggestion
	 * @param newStacks   The stacks to add. Only {@linkplain #isEnabled(ItemStack) enabled} stacks will be added.
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	/**
+	 * Adds stacks before an existing stack to the group, or at the end, if the stack isn't in the group.
+	 *
+	 * @param beforeFirst Add <code>newStacks</code> before the first group entry matching this stack (compared using {@link ItemStack#canCombine}).
```
```suggestion
	 * @param beforeFirst Add {@code newStacks} before the first group entry matching this stack (compared using {@link ItemStack#canCombine}).
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	 * Adds stacks before an existing stack to the group, or at the end, if the stack isn't in the group.
+	 *
+	 * @param beforeFirst Add <code>newStacks</code> before the first group entry matching this stack (compared using {@link ItemStack#canCombine}).
+	 * @param newStacks   The stacks to add. Only {@link #isEnabled(ItemStack) enabled} stacks will be added.
```
```suggestion
	 * @param newStacks   The stacks to add. Only {@linkplain #isEnabled(ItemStack) enabled} stacks will be added.
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	/**
+	 * Adds stacks before the first group entry matching a predicate, or at the end, if no entries match.
+	 *
+	 * @param beforeFirst Add <code>newStacks</code> before the first group entry matching this predicate.
```
```suggestion
	 * @param beforeFirst Add {@code newStacks} before the first group entry matching this predicate.
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	 * Adds stacks before the first group entry matching a predicate, or at the end, if no entries match.
+	 *
+	 * @param beforeFirst Add <code>newStacks</code> before the first group entry matching this predicate.
+	 * @param newStacks   The stacks to add. Only {@link #isEnabled(ItemStack) enabled} stacks will be added.
```
```suggestion
	 * @param newStacks   The stacks to add. Only {@linkplain #isEnabled(ItemStack) enabled} stacks will be added.
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/IdentifiableItemGroup.java:
```diff
@@ -14,21 +14,19 @@
+public interface IdentifiableItemGroup {
 	/**
-	 * @return A set of all sprites that should be loaded before this sprite.
+	 * @return The unique identifier of this {@link ItemGroup}.
```
```suggestion
	 * {@return the unique identifier of this {@link ItemGroup}}
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/mixin/itemgroup/ItemGroupAccessor.java:
```diff
@@ -14,16 +14,17 @@
-	ResourcePackSource getResourcePackSource();
+@Mixin(ItemGroup.class)
+public interface ItemGroupAccessor {
+	@Accessor("index")
```
```suggestion
	@Accessor
```
The field name is not needed since it matches the setter

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/mixin/itemgroup/ItemGroupsAccessor.java:
```diff
@@ -14,29 +14,28 @@
+	}
 
-		System.arraycopy(tempGroups, 0, GROUPS, 0, tempGroups.length);
+	@Invoker("asArray")
```
```suggestion
	@Invoker
```

Juuxel commented on fabric-item-group-api-v1/src/client/resources/fabric-item-groups-v1.client.mixins.json:
```diff
@@ -1,6 +1,6 @@
 {
```
This file should be called fabric-item-group-api-v1.client.mixins.json

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/mixin/itemgroup/ItemGroupsMixin.java:
```diff
@@ -0,0 +1,38 @@
+public class ItemGroupsMixin {
+	@Inject(method = "asArray", at = @At("HEAD"))
+	private static void asArray(ItemGroup[] groups, CallbackInfoReturnable<ItemGroup[]> cir) {
+		// Ensure that all item groups have a none null id.
```
```suggestion
		// Ensure that all item groups have a nonnull id.
```
(or "non-null" but javax annotations uses this spelling)

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/mixin/itemgroup/ItemGroupsMixin.java:
```diff
@@ -0,0 +1,38 @@
+		// Ensure that all item groups have a none null id.
+		for (ItemGroup group : groups) {
+			if (group.getId() == null) {
+				throw new NullPointerException("Item group %s has a null id".formatted(group.getClass()));
```
```suggestion
				throw new NullPointerException("Item group %s has a null id".formatted(group.getClass().getName()));
```
"Item group class mymod.MyItemGroup has..." is a bit ugly

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/mixin/itemgroup/ItemGroupsMixin.java:
```diff
@@ -0,0 +1,38 @@
+import net.minecraft.item.ItemGroups;
+
+@Mixin(ItemGroups.class)
+public class ItemGroupsMixin {
```
```suggestion
abstract class ItemGroupsMixin {
```
Mixins should be package-private according to fabric conventions

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/mixin/itemgroup/ItemGroupMixin.java:
```diff
@@ -0,0 +1,95 @@
+import net.fabricmc.fabric.impl.itemgroup.MinecraftItemGroups;
+
+@Mixin(ItemGroup.class)
+public class ItemGroupMixin implements IdentifiableItemGroup {
```
```suggestion
abstract class ItemGroupMixin implements IdentifiableItemGroup {
```

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/impl/itemgroup/MinecraftItemGroups.java:
```diff
@@ -0,0 +1,60 @@
+	public static final Identifier SPAWN_EGGS_ID = new Identifier("minecraft:spawn_eggs");
+	public static final Identifier INVENTORY_ID = new Identifier("minecraft:inventory");
+
+	public static final Map<ItemGroup, Identifier> MAP = new ImmutableMap.Builder<ItemGroup, Identifier>()
```
Maybe something more descriptive like `GROUP_ID_MAP`?

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/impl/itemgroup/ItemGroupHelper.java:
```diff
@@ -0,0 +1,50 @@
+		}
+
+		final int index = ItemGroups.GROUPS.length;
+		final ItemGroup[] itemGroups = Stream.concat(Arrays.stream(ItemGroups.GROUPS), Stream.of(itemGroup))
+				.toArray(ItemGroup[]::new);
```
Don't we have the apache commons lib that has a util method for adding an element to an array? Doesn't really matter though tbh

modmuss50 replied:
ArrayUtils seems to do the job, thanks 👍

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/ItemGroupEvents.java:
```diff
@@ -0,0 +1,59 @@
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.fabricmc.fabric.impl.itemgroup.ItemGroupEventsImpl;
+
+public class ItemGroupEvents {
```
```suggestion
public final class ItemGroupEvents {
```
Should also have a private constructor, and it needs class jd

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/ItemGroupEvents.java:
```diff
@@ -0,0 +1,59 @@
+		return modifyEntriesEvent(itemGroup.getId());
+	}
+
+	public static Event<ModifyEntries> modifyEntriesEvent(Identifier identifier) {
+		return ItemGroupEventsImpl.getOrCreateModifyEntriesEvent(identifier);
```
```suggestion
	public static Event<ModifyEntries> modifyEntriesEvent(Identifier groupId) {
		return ItemGroupEventsImpl.getOrCreateModifyEntriesEvent(groupId);
```
Just "identifier" doesn't really describe *what* it identifies

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/ItemGroupEvents.java:
```diff
@@ -0,0 +1,59 @@
+		return ItemGroupEventsImpl.getOrCreateModifyEntriesEvent(identifier);
+	}
+
+	@FunctionalInterface
+	public interface ModifyEntries {
+		void modifyEntries(FeatureSet featureSet, FabricItemGroupEntries entries);
+	}
+
+	@FunctionalInterface
+	public interface ModifyEntriesAll {
+		void modifyEntries(ItemGroup group, FeatureSet featureSet, FabricItemGroupEntries entries);
+	}
```
Do we really need two separate interfaces for this? IMO `ModifyEntries` could just have the group param

shartte replied:
I thought about it, but then all of your event listeners that target a specific group will have a useless 1st parameter

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/IdentifiableItemGroup.java:
```diff
@@ -14,21 +14,19 @@
- * that should be processed prior to this sprite.
- *
- * <p>Best used in conjunction with {@link net.fabricmc.fabric.api.event.client.ClientSpriteRegistryCallback}.
+ * This interface is automatically implemented on all {@link ItemGroup} via Mixin and interface injection.
```
```suggestion
 * This interface is automatically injected to {@link ItemGroup}.
```
This is how it's done in existing code (loot api), mixins are an impl detail

Juuxel commented on fabric-item-group-api-v1/src/testmod/resources/fabric.mod.json:
```diff
@@ -1,12 +1,12 @@
   "schemaVersion": 1,
-  "id": "fabric-item-groups-v0-testmod",
-  "name": "Fabric Item Groups (v0) Test Mod",
+  "id": "fabric-item-groups-v1-testmod",
+  "name": "Fabric Item Groups (v1) Test Mod",
```
```suggestion
  "id": "fabric-item-group-api-v1-testmod",
  "name": "Fabric Item Group API (v1) Test Mod",
```
The mod id in the code is already correct btw

modmuss50 replied:
The mixin config name is also wrong, will all be fixed in my next commit.

AlphaMode commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/mixin/itemgroup/ItemGroupMixin.java:
```diff
@@ -41,13 +45,42 @@ public class ItemGroupMixin implements IdentifiableItemGroup {
 
 		if (modifyEntriesEvent != null) {
-			modifyEntriesEvent.invoker().modifyItems(featureSet, entries);
+			modifyEntriesEvent.invoker().modifyContent(featureSet, content);
```
Why aren't we passing the search field here couldn't that be useful for modders?

shartte replied:
No it wouldn't be useful. The modified lists are stored/cached in the ItemGroup and used in both queries with search=true and search=false.

modmuss50 committed:
Update fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 committed:
Update fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 committed:
Update fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 committed:
Update fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 committed:
Update fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 committed:
Apply suggestions from code review

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 committed:
Review comments, more javadoc

modmuss50 committed:
Work around javadoc generation errors.

modmuss50 committed:
Apply suggestions from code review

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

modmuss50 committed:
Enable javadoc failOnError

Juuxel commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/IdentifiableItemGroup.java:
```diff
@@ -24,7 +24,7 @@
 public interface IdentifiableItemGroup {
 	/**
-	 * {@return the unique identifier of this {@link ItemGroup}}
+	 * {@return the unique identifier of this {@link ItemGroup}}.
```
Does this lead to an extra dot at the end?

modmuss50 replied:
Checkstyle fails without it.

Juuxel replied:
I know but that's not what I was asking 😛

modmuss50 replied:
I should learn to read one day...

Just checked, no it doesnt.

Juuxel replied:
TIL, then it's all good.

apple502j commented on fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricBuiltinRegistriesProvider.java:
```diff
@@ -0,0 +1,113 @@
+		);
+	}
+
+	public void run(DataWriter writer) {
```
Missing `@Override`?

apple502j commented on fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricBuiltinRegistriesProvider.java:
```diff
@@ -0,0 +1,113 @@
+		}
+	}
+
+	public String getName() {
```
Same here.

modmuss50 committed:
Update fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java

Co-authored-by: Juuz <6596629+Juuxel@users.noreply.github.com>

apple502j commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroup.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Extend this abstract class to create a new {@link ItemGroup} for a mod.
+ *
+ * <p>Each new instance of this class is automatically appended to {@link ItemGroups#GROUPS}
```
```suggestion
 * <p>Each new instance of this class is automatically appended to {@link ItemGroups#GROUPS}.
```

apple502j commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	/**
+	 * @return The currently enabled feature set.
+	 */
+	public FeatureSet getFeatureSet() {
```
```suggestion
	public FeatureSet getEnabledFeatures() {
```

modmuss50 replied:
This seems correct to me.

modmuss50 replied:
2nd thoughts this seems to make more sense.

apple502j commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/FabricItemGroupEntries.java:
```diff
@@ -0,0 +1,464 @@
+	 * @return True if the item of a given stack is enabled in the current {@link FeatureSet}.
+	 * @see Item#method_45382
+	 */
+	public boolean isEnabled(ItemStack stack) {
```
Why is this a public API?

apple502j commented on fabric-item-group-api-v1/src/main/java/net/fabricmc/fabric/api/itemgroup/v1/ItemGroupEvents.java:
```diff
@@ -0,0 +1,68 @@
+
+	@FunctionalInterface
+	public interface ModifyEntries {
+		void modifyEntries(FeatureSet featureSet, FabricItemGroupEntries entries);
```
What is this `featureSet`?

modmuss50 replied:
Good spot, Ill remove this. I dont think many mods will make use of this, and if they do they can easily get it from FabricItemGroupEntries.

apple502j commented on fabric-item-group-api-v1/src/main/resources/fabric-item-group-api-v1.mixins.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "required": true,
+  "package": "net.fabricmc.fabric.mixin.itemgroup",
+  "compatibilityLevel": "JAVA_16",
```
```suggestion
  "compatibilityLevel": "JAVA_17",
```

apple502j commented on fabric-item-group-api-v1/src/client/resources/fabric-item-group-api-v1.client.mixins.json:
```diff
@@ -1,6 +1,6 @@
   "required": true,
-  "package": "net.fabricmc.fabric.mixin.item.group.client",
+  "package": "net.fabricmc.fabric.mixin.itemgroup.client",
   "compatibilityLevel": "JAVA_16",
```
```suggestion
  "compatibilityLevel": "JAVA_17",
```

modmuss50 committed:
Review feedback

modmuss50 committed:
featureSet -> enabledFeatures

modmuss50 committed:
Remove enabled features param from modify events.

modmuss50 committed:
Minor bump
